agent/cache/entry.go

```diff
@@ -26,9 +26,10 @@ type cacheEntry struct {
 	Index uint64
 
 	// Metadata that is used for internal accounting
-	Valid    bool          // True if the Value is set
-	Fetching bool          // True if a fetch is already active
-	Waiter   chan struct{} // Closed when this entry is invalidated
+	Valid        bool          // True if the Value is set
+	Fetching     bool          // True if a fetch RPC is already active
+	HasGoroutine bool          // True if a fetch goroutine is running.
```

Comment:
That's actually the bug being fixed here.

The code pre-PR slightly conflated two scenarios:

1. `fetching` means there is a blocking query that has not returned yet
2. `fetching` means that there is a goroutine that _may_ be executing an RPC right now

Within the body of that goroutine, when the RPC returns it updated the fetching value to false which _breaks_ code that assumes (2). But we need to do that so other code can use the existence of a blocked RPC to imply:
```
			// There is an active blocking query for this data, which has not
			// returned. We can logically deduce that the contents of the cache
			// are actually current, and we can simply return this while
			// leaving the blocking query alone.
			return true, true, entry
```

So we sorta need to know (1) for the above cache freshness inference and (2) to avoid the goroutine leak.

Reply:
Thank you for explaining this.

I see it now. The nuance is between: I have a blocking query for it (which mean what I have is up to date) and I'm trying to update the cache but nothing is in there for the moment or I have some error.